A79-non-per-call-metrics-architecture.md

```diff
@@ -0,0 +1,432 @@
+   absl::string_view authority() const { return authority_; }
+
+  private:
+   absl::string_view target_;
+   absl::string_view authority_;
+ };
+
+ virtual bool IsEnabledForChannel(const ChannelScope& scope);
+ virtual bool IsEnabledForServer(ChannelArgs& args);
+
+ virtual void AddCounter(
+     GlobalInstrumentsRegistry::GlobalUInt64CounterHandle handle,
+     uint64_t value, absl::Span<const absl::string_view> label_values,
+     absl::Span<const absl::string_view> optional_values) = 0;
+ virtual void RecordHistogram(
+     GlobalInstrumentsRegistry::GlobalDoubleHistogramHandle handle,
+     double value, absl::Span<const absl::string_view> label_values,
+     absl::Span<const absl::string_view> optional_values) = 0;
+};
+
+class GlobalStatsPluginRegistry {
+ public:
+  class StatsPluginGroup {
+   public:
+    // Use the stats plugin group object to record metrics
+    template <class T, class U>
+    void AddCounter(T handle, U value,
+                    absl::Span<const absl::string_view> label_values,
+                    absl::Span<const absl::string_view> optional_values) {
+      for (auto& plugin : plugins_) {
+        plugin->AddCounter(handle, value, label_values, optional_values);
+      }
+    }
+    template <class T, class U>
+    void RecordHistogram(T handle, U value,
+                         absl::Span<const absl::string_view> label_values,
+                         absl::Span<const absl::string_view> optional_values) {
+      for (auto& plugin : plugins_) {
+        plugin->RecordHistogram(handle, value, label_values, optional_values);
+      }
+    }
+
+    // Use the stats plugin to get a representation of label values that can be
+    // saved for multiple uses later.
+    RefCountedPtr<LabelValueSet> MakeLabelValueSet(
+        absl::Span<const absl::string_view> label_values);
+  };
+
+  static void RegisterStatsPlugin(std::shared_ptr<StatsPlugin> plugin);
+
+  // Invoked to get a stats plugin list for a specified scope.
+  static StatsPluginGroup GetStatsPluginsForChannel(ChannelScope scope);
+
+ private:
+  static NoDestruct<std::vector<std::shared_ptr<StatsPlugin>>> plugins_;
+};
+```
+
+##### Java
+
+(To be filled)
+
+##### Go
+
+(To be filled)
+
+### Metric Instrument Naming Conventions
+
+gRPC metrics follow the same general
+[guidelines](https://opentelemetry.io/docs/specs/semconv/general/metrics/#general-guidelines)
+as OpenTelemetry. The following additional guidelines apply -
+
+*   Metrics defined by gRPC should go under the `grpc.` namespace.
+*   Third-party plugins to gRPC should NOT use the `grpc.` namespace.
+*   Newer metrics should follow the same style as existing metrics.
+*   Metrics from load-balancing policies should be nested under `grpc.lb.`, e.g.
+    `grpc.lb.pick_first.`.
+*   Metrics from resolvers should be nested under `grpc.resolver.`, e.g.
+    `grpc.resolver.dns.`.
+*   Metrics from transports should be nested under `grpc.transport.`, e.g.,
+    `grpc.transport.http2.`
```

Comment:
Clearly by this you mean "Metrics from first-party gRPC ___s".  Any guidance for third-party plugins for these components?

Reply:
The only recommendation from gRPC for third-party plugins is to NOT use the `grpc.` namespace. In general, the same recommendations from OpenTelemetry (https://opentelemetry.io/docs/specs/semconv/general/metrics/#general-guidelines) should be followed